vms/platformvm/state/state.go

```diff
@@ -944,74 +966,168 @@ func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
 	return nil
 }
 
-func (s *state) GetValidatorWeightDiffs(height uint64, subnetID ids.ID) (map[ids.NodeID]*ValidatorWeightDiff, error) {
-	prefixStruct := heightWithSubnet{
-		Height:   height,
-		SubnetID: subnetID,
-	}
-	prefixBytes, err := blocks.GenesisCodec.Marshal(blocks.Version, prefixStruct)
-	if err != nil {
-		return nil, err
-	}
-	prefixStr := string(prefixBytes)
+func (s *state) ApplyValidatorWeightDiffs(
+	ctx context.Context,
+	validators map[ids.NodeID]*validators.GetValidatorOutput,
+	startHeight uint64,
+	endHeight uint64,
+	subnetID ids.ID,
+) error {
+	diffIter := s.flatValidatorWeightDiffsDB.NewIteratorWithStartAndPrefix(
+		getStartWeightKey(subnetID, startHeight),
+		subnetID[:],
+	)
+	defer diffIter.Release()
 
-	if weightDiffs, ok := s.validatorWeightDiffsCache.Get(prefixStr); ok {
-		return weightDiffs, nil
-	}
+	prevHeight := startHeight + 1
```

Comment:
nit: I don't love this +1 then -1 thing to `prevHeight` but I don't think it is wrong

Reply:
I prefer `nextHeight` iteration but definitely a nit

Reply:
I decided to keep `prevHeight` - because it _is_ the previous height during the iteration... Whereas `nextHeight` would not necessarily be the next height observed in the iteration.